pkg/resource/user_group/custom_update_api.go

```diff
@@ -0,0 +1,185 @@
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package user_group
+
+import (
+	"context"
+	"errors"
+	svcapitypes "github.com/aws-controllers-k8s/elasticache-controller/apis/v1alpha1"
+	ackv1alpha1 "github.com/aws-controllers-k8s/runtime/apis/core/v1alpha1"
+	ackcompare "github.com/aws-controllers-k8s/runtime/pkg/compare"
+	"github.com/aws-controllers-k8s/runtime/pkg/requeue"
+	svcsdk "github.com/aws/aws-sdk-go/service/elasticache"
+)
+
+// Implements custom logic for UpdateUserGroup
+func (rm *resourceManager) customUpdateUserGroup(
+	ctx context.Context,
+	desired *resource,
+	latest *resource,
+	delta *ackcompare.Delta,
+) (*resource, error) {
+
+	// Check for the user group status
+	if latest.ko.Status.Status == nil || *latest.ko.Status.Status != "active" {
+		return nil, requeue.NeededAfter(
+			errors.New("user group can not be modified, it is not in 'active' state"),
+			requeue.DefaultRequeueAfterDuration)
+	}
+
+	for _, diff := range delta.Differences {
+		if diff.Path.Contains("Spec.UserIDs") {
+			existingUserIdsMap := rm.createMapForUserIds(diff.B.([]*string))
+			requiredUserIdsMap := rm.createMapForUserIds(diff.A.([]*string))
+
+			// If a user ID is not required to be deleted or added set its value as false
+			for userId, _ := range existingUserIdsMap {
+				if _, ok := requiredUserIdsMap[userId]; ok {
+					requiredUserIdsMap[userId] = false
+					existingUserIdsMap[userId] = false
+				}
+			}
+
+			input, err := rm.newUpdateRequestPayload(ctx, desired)
+
+			if err != nil {
+				return nil, err
+			}
+
+			// User Ids to add
+			{
+				var userIdsToAdd []*string
+
+				for userId, include := range requiredUserIdsMap {
+					if include {
+						userIdsToAdd = append(userIdsToAdd, &userId)
+					}
+				}
+
+				input.SetUserIdsToAdd(userIdsToAdd)
+			}
+
+			// User Ids to remove
+			{
+				var userIdsToRemove []*string
+
+				for userId, include := range existingUserIdsMap {
+					if include {
+						userIdsToRemove = append(userIdsToRemove, &userId)
+					}
+				}
+
+				input.SetUserIdsToRemove(userIdsToRemove)
+			}
+
+			if len(input.UserIdsToAdd) == 0 && len(input.UserIdsToRemove) == 0 {
+				rm.customSetOutput(desired.ko.Spec.UserIDs, desired.ko.Spec.Engine,
+					latest.ko.Status.Status, latest.ko)
+				return &resource{latest.ko}, nil
+			}
+
+			resp, respErr := rm.sdkapi.ModifyUserGroupWithContext(ctx, input)
+			rm.metrics.RecordAPICall("UPDATE", "ModifyUserGroup", respErr)
+			if respErr != nil {
+				return nil, respErr
+			}
+			// Merge in the information we read from the API call above to the copy of
+			// the original Kubernetes object we passed to the function
+			ko := desired.ko.DeepCopy()
```

Comment:
Is the below block of code same as the generated code in `sdk.go`? If so, we may be able to avoid doing this by placing the above code where you add/remove Users in a `post_build_request` hook, and then just defer to the generated code afterwards

Reply:
Generated code did not seem to be concise and have additional statements which did not seem to be required thus I created a completely custom update method. I would keep this same for now will again revisit later to see if this can be improved.

Reply:
Sounds good. Before, I was aware of `custom_implementation` in the generator config but didn't know you could replace the entire `sdkUpdate` using `update_operation.custom_method_name`.